File: e_mart/e_mart/custom_scripts/sales_invoice/sales_invoice.js
// Copyright (c) 2025, efeone and contributors
// For license information, please see license.txt

frappe.ui.form.on('Sales Invoice', {
	onload(frm) {
		frm.fields_dict.buyback_items.grid.wrapper.on('change', function () {
			update_total_buyback_amount(frm);
			update_rounded_total(frm);
		});
		frm.set_query('item', 'buyback_items', function(doc, cdt, cdn) {
			return {
				filters: {
					buyback_item: 1
				}
			};
		});
		set_finance_filter(frm);
		attach_sales_expense_grid_events(frm);
		calculate_total_expense(frm);
		calculate_total_down_payment(frm);
		update_down_payment_checkbox(frm);
		calculate_total_emi_amount(frm);
	},
	refresh: function (frm) {
		setTimeout(() => {
			frm.remove_custom_button('Payment', 'Create');
			if (frm.doc.docstatus === 1 && flt(frm.doc.outstanding_amount) > 0) {
				frm.add_custom_button(__('Payment'), function() {
					show_payment_popup(frm);
				}, __('Create'));
			}
		}, 100);
		if (frm.doc.docstatus === 0) {
			update_outstanding_amount(frm);
			update_rounded_total(frm);
		}
		if (!frm.is_new() && frm.doc.sales_type === "EMI") {
			frm.add_custom_button(__('Create Finance Invoice'), function () {
				frappe.call({
					method: "e_mart.e_mart.custom_scripts.sales_invoice.sales_invoice.create_finance_invoice",
					args: {
						sales_invoice_name: frm.doc.name
					},
					callback: function (r) {
						if (r.message) {
							frappe.set_route('Form', 'Finance Invoice', r.message);
						}
					}
				});
			});
			if (frm.doc.docstatus === 1 && frm.doc.down_payment && frm.doc.down_payment_amount && frm.doc.down_payment_paid == 0 ) {
				frm.add_custom_button(__('Down Payment'), function() {
					frappe.model.open_mapped_doc({
						method: "e_mart.e_mart.custom_scripts.sales_invoice.sales_invoice.make_down_payment_entry",
						source_name: frm.doc.name
					});
				}, __('Create'));
			}
		}
		calculate_total_expense(frm);
		calculate_total_down_payment(frm);
		update_down_payment_checkbox(frm);
		calculate_total_emi_amount(frm);
	},
	sales_type(frm) {
		set_finance_filter(frm);
		sync_sales_type_to_items(frm);
	},
	total: function(frm) {
		update_outstanding_amount(frm);
		update_rounded_total(frm);
	},
	total_taxes_and_charges(frm) {
		update_outstanding_amount(frm);
		update_rounded_total(frm);
	},

	buyback_amount(frm) {
		update_outstanding_amount(frm);
		update_rounded_total(frm);
	},

	is_buyback(frm) {
		update_outstanding_amount(frm);
		update_rounded_total(frm);
	},
	validate(frm) {
		calculate_total_expense(frm);
		update_profit_for_commission(frm);
		calculate_total_down_payment(frm);
		calculate_total_emi_amount(frm);
	},
	calculate_totals(frm) {
		setTimeout(() => {
			update_outstanding_amount(frm);
			update_rounded_total(frm);
			calculate_total_expense(frm);
			calculate_total_down_payment(frm);
			calculate_total_emi_amount(frm);
		}, 200);
	},
	total_expense(frm) {
		update_profit_for_commission(frm);
	}
});
// Buyback Item child table
frappe.ui.form.on('Buyback Item', {
	qty(frm, cdt, cdn) {
		calculate_row_amount(frm, cdt, cdn);
	},
	rate(frm, cdt, cdn) {
		calculate_row_amount(frm, cdt, cdn);
	},
	buyback_items_add(frm) {
		update_total_buyback_amount(frm);
		update_grand_total(frm);
		update_rounded_total(frm);
		update_outstanding_amount(frm)
	},
	buyback_items_remove(frm) {
		update_total_buyback_amount(frm);
		update_grand_total(frm);
		update_rounded_total(frm);
		update_outstanding_amount(frm)
	}
});
// Sales Invoice Items child table
frappe.ui.form.on('Sales Invoice Item', {
	qty(frm, cdt, cdn) {
		frm.trigger("calculate_totals");
		update_emi_amount(frm, cdt, cdn);
		set_valuation_and_gross_profit(frm, cdt, cdn);
	},
	rate(frm, cdt, cdn) {
		frm.trigger("calculate_totals");
		update_emi_amount(frm, cdt, cdn);
		set_valuation_and_gross_profit(frm, cdt, cdn);
	},
	amount(frm, cdt, cdn) {
		frm.trigger("calculate_totals");
		calculate_total_expense(frm);
		update_emi_amount(frm, cdt, cdn);
	},
	down_payment(frm, cdt, cdn) {
		update_emi_amount(frm, cdt, cdn);
		calculate_total_down_payment(frm);
	},
	is_down_payment(frm, cdt, cdn) {
		update_down_payment_checkbox(frm);
	},
	emi_amount(frm, cdt, cdn) {
		calculate_total_emi_amount(frm);
	},
	items_add(frm) {
		frm.trigger("calculate_totals");
		update_down_payment_checkbox(frm);
	},
	items_remove(frm) {
		frm.trigger("calculate_totals");
		calculate_total_down_payment(frm);
		update_down_payment_checkbox(frm);
		calculate_total_emi_amount(frm);
	},
	item_code(frm, cdt, cdn) {
		let row = locals[cdt][cdn];
		if (row.item_code) {
			frappe.db.get_value("Item", row.item_code, "sales_expense_contribution")
			.then(r => {
				if (r.message) {
					frappe.model.set_value(cdt, cdn, "sales_expense_contribution", r.message.sales_expense_contribution);
					update_profit_for_commission(frm);
					}
				});
			}
			set_valuation_and_gross_profit(frm, cdt, cdn);
	},
	sales_expense_contribution(frm, cdt, cdn) {
		update_profit_for_commission(frm);
	},
	form_render: function(frm, cdt, cdn) {
		let row = locals[cdt][cdn];
		row.sales_type = frm.doc.sales_type;
		refresh_field("items");
	}
});

// Taxes child table
frappe.ui.form.on('Sales Taxes and Charges', {
	tax_amount(frm) {
		frm.trigger("calculate_totals");
	},
	rate(frm) {
		frm.trigger("calculate_totals");
	},
	taxes_add(frm) {
		frm.trigger("calculate_totals");
	},
	taxes_remove(frm) {
		frm.trigger("calculate_totals");
	}
});

// Calculate Buyback Row Amount
function calculate_row_amount(frm, cdt, cdn) {
	let row = locals[cdt][cdn];
	row.amount = (flt(row.qty) || 0) * (flt(row.rate) || 0);
	frm.fields_dict.buyback_items.grid.refresh();
	update_total_buyback_amount(frm);
}

// Sum all row amounts from buyback_items_amount
function update_total_buyback_amount(frm) {
	let total = 0;
	(frm.doc.buyback_items || []).forEach(row => {
		total += flt(row.amount || 0);
	});
	frm.set_value('buyback_amount', total);
	update_outstanding_amount(frm);
	update_rounded_total(frm);
	update_grand_total(frm);
}

/*
* Calculate Outstanding Amount based on buyback amount
*/
function update_outstanding_amount(frm) {
	if (frm.doc.docstatus === 0) { 
		const total = flt(frm.doc.total || 0);
		const taxes = flt(frm.doc.total_taxes_and_charges || 0);
		const buyback = flt(frm.doc.buyback_amount || 0);

		let outstanding = total + taxes;
		if (frm.doc.is_buyback) {
			outstanding -= buyback;
		}
		frm.set_value('outstanding_amount', Math.max(outstanding, 0));
		update_grand_total(frm); 
		update_emi_amount(frm);
	}
}

/*
* Calculate Rounded Total based on buyback amount
*/
function update_rounded_total(frm) {
	const total = flt(frm.doc.total || 0);
	const taxes = flt(frm.doc.total_taxes_and_charges || 0);
	const buyback = flt(frm.doc.buyback_amount || 0);

	let grand_total = total + taxes;
	if (frm.doc.is_buyback) {
		grand_total -= buyback;
	}
	frm.set_value('rounded_total', Math.round(grand_total));
}

/*
* Calculate Grand Total based on buyback amount
*/
function update_grand_total(frm) {
	const total = flt(frm.doc.total || 0);
	const taxes = flt(frm.doc.total_taxes_and_charges || 0);
	const buyback = flt(frm.doc.buyback_amount || 0);

	let grand_total = total + taxes;
	if (frm.doc.is_buyback) {
		grand_total -= buyback;
	}
	frm.set_value('grand_total', grand_total);
}

// Filter customer field to show only providers when sales_type is EMI
// If selected customer is not a provider, clear it
function set_finance_filter(frm) {
	frm.set_query("mode_of_payment", () => {
		if (frm.doc.sales_type === "EMI") {
			return {
				filters: {
					is_finance: 1
				}
			};
		} else {
			return {};
		}
	});

	frm.set_query("emi_provider", () => {
		if (frm.doc.sales_type === "EMI") {
			return {
				filters: {
					is_provider: 1
				}
			};
		} else {
			return {};
		}
	});
}

/**
 * Calculates the EMI amount after deducting the down payment
 * and updates the emi_amount field immediately.
 */
function update_emi_amount(frm, cdt, cdn) {
	const row = locals[cdt][cdn];
	if (frm.doc.docstatus !== 0) {
		return;
	}
	const down_payment = flt(row.down_payment || 0);
	const total_amount = flt(row.amount || 0);
	const emi_amount = total_amount - down_payment;

	frappe.model.set_value(cdt, cdn, 'emi_amount', Math.max(emi_amount, 0));
}

frappe.ui.form.on('Sales Expenses', {
	amount(frm, cdt, cdn) {
		calculate_total_expense(frm);
	},
	sales_expenses_add(frm) {
		calculate_total_expense(frm);
	},
	sales_expenses_remove(frm) {
		calculate_total_expense(frm);
	}
});

/**
 * Attach grid-remove-row event to Sales Expenses child table.
 */
function attach_sales_expense_grid_events(frm) {
	if (frm._sales_expense_grid_attached) return;

	frm.fields_dict['sales_expenses'].grid.wrapper.on('grid-remove-row', () => {
		calculate_total_expense(frm);
	});

	frm._sales_expense_grid_attached = true;
}

/**
 * Sum up all amounts in Sales Expenses child table
 * and update total_expense field.
 */
function calculate_total_expense(frm) {
	let total = 0;

	(frm.doc.sales_expenses || []).forEach(row => {
		total += flt(row.amount || 0);
	});
	frm.set_value('total_expense', total);
	update_profit_for_commission(frm);
}

frappe.ui.form.on("Sales Team", {
	allocated_percentage: function(frm, cdt, cdn) {
			var sales_person = frappe.get_doc(cdt, cdn);
			let row = locals[cdt][cdn];
			if (sales_person.allocated_percentage) {
				sales_person.allocated_percentage = sales_person.allocated_percentage;
				sales_person.total_commission_rate = frm.doc.total_commission_rate;
				sales_person.incentive = flt(
					(sales_person.total_commission_rate * row.allocated_percentage) / 100.0,
				);
				frm.refresh_field('sales_team');
			}
	}
});

/**
 * Calculates the profit for Commission for each item by subtracting its Sales Expense Contribution
 * from the total expense.
 */
function update_profit_for_commission(frm) {
	const total_expense = flt(frm.doc.total_expense || 0);

	(frm.doc.items || []).forEach(row => {
		const contribution = flt(row.sales_expense_contribution || 0);
		const profit = total_expense - contribution;
		// Set profit_for_commission for each item
		frappe.model.set_value(row.doctype, row.name, 'profit_for_commission', profit);
	});

	frm.refresh_field('items');
}

/**
 * Synchronizes the 'sales_type' field from the main Sales Invoice
 * to all child rows in the items table
 */
function sync_sales_type_to_items(frm) {
	(frm.doc.items || []).forEach(row => {
		row.sales_type = frm.doc.sales_type;
	});
	frm.refresh_field("items");
}

/**
 * Calculates and sets the total down payment amount on the invoice
 * by summing up all 'down_payment' values from the items table.
 */
function calculate_total_down_payment(frm) {
	let total = 0;
	(frm.doc.items || []).forEach(item => {
		if (item.down_payment) {
			total += item.down_payment;
		}
	});
	frm.set_value('down_payment_amount', total);
}

/**
 * Updates the 'down_payment' checkbox on the Sales Invoice.
 * It is set to true (1) if any item row has 'is_down_payment' checked.
 */
function update_down_payment_checkbox(frm) {
	let has_down_payment = (frm.doc.items || []).some(item => item.is_down_payment);
	frm.set_value('down_payment', has_down_payment ? 1 : 0);
}

/**
 * Calculates and sets the total emi amount on the invoice
 * by summing up all 'emi_amount' values from the items table.
 */
function calculate_total_emi_amount(frm) {
	let total = 0;
	(frm.doc.items || []).forEach(item => {
		if (item.emi_amount) {
			total += item.emi_amount;
		}
	});
	frm.set_value('emi_amount', total);
}
/**
 * Fetches the valuation rate from the 'Bin' table based on the item and warehouse,
 * and calculates gross profit per item. 
 */
function set_valuation_and_gross_profit(frm, cdt, cdn) {
	const row = locals[cdt][cdn];

	if (row.item_code && row.warehouse) {
		frappe.call({
			method: 'frappe.client.get_value',
			args: {
				doctype: 'Bin',
				filters: {
					item_code: row.item_code,
					warehouse: row.warehouse
				},
				fieldname: ['valuation_rate']
			},
			callback: function(r) {
				if (r.message) {
					const valuation_rate = flt(r.message.valuation_rate);
					frappe.model.set_value(cdt, cdn, 'valuation_rate', valuation_rate);

					if (row.rate) {
						const gross_profit = (row.rate - valuation_rate) * row.qty;
						frappe.model.set_value(cdt, cdn, 'gross_profit', gross_profit);
					}
				}
			}
		});
	}
}
/**
 * Opens a popup to record multiple payments for the Sales Invoice.
 * Dynamically updates balance and creates a Payment Entry on submission.
 */
function show_payment_popup(frm) {
	let outstanding_amount = frm.doc.outstanding_amount || 0;

	const dialog = new frappe.ui.Dialog({
		title: 'Record Payment',
		fields: [
			{
				fieldtype: 'HTML',
				fieldname: 'amount_summary_display',
				options: `
					<div style="display: flex; justify-content: space-between; gap: 20px; margin-bottom: 10px;">
						<div id="outstanding_amount_box" style="font-size: 16px; font-weight: bold;">
							Outstanding Amount: ₹${outstanding_amount.toFixed(2)}
						</div>
						<div id="balance_to_pay_display" style="font-size: 16px; font-weight: bold;">
							Balance to Pay: ₹${outstanding_amount.toFixed(2)}
						</div>
					</div>`
			},
			{
				fieldname: 'payments',
				fieldtype: 'Table',
				label: 'Payments',
				cannot_add_rows: false,
				in_place_edit: true,
				data: [],
				get_data: () => {
					return dialog.get_value('payments') || [];
				},
				fields: [
					{
						fieldname: 'mode_of_payment',
						fieldtype: 'Link',
						label: 'Mode of Payment',
						options: 'Mode of Payment',
						in_list_view: 1,
						reqd: 1
					},
					{
						fieldname: 'reference_no',
						fieldtype: 'Data',
						label: 'Reference No',
						in_list_view: 1
					},
					{
						fieldname: 'reference_date',
						fieldtype: 'Date',
						label: 'Reference Date',
						in_list_view: 1
					},
					{
						fieldname: 'amount',
						fieldtype: 'Currency',
						label: 'Amount',
						in_list_view: 1,
						reqd: 1
					}
				]
			}
		],
		primary_action_label: 'Submit Payment',
		primary_action(values) {
			const payments = values.payments || [];
			if (!payments.length) {
				frappe.msgprint(__('Please add at least one payment'));
				return;
			}

			frappe.call({
				method: 'e_mart.e_mart.custom_scripts.sales_invoice.sales_invoice.make_payment_entry',
				args: {
					sales_invoice: frm.doc.name,
					payments: payments
				},
				callback: function (r) {
					if (r.message) {
						frappe.msgprint('Payment Entry Created: ' + r.message);
						dialog.hide();
						frm.reload_doc();
					}
				}
			});
		}
	});

	dialog.show();
	const update_balance = () => {
		const data = dialog.get_value('payments') || [];
		const total_paid = data.reduce((acc, row) => acc + (parseFloat(row.amount) || 0), 0);
		const balance = outstanding_amount - total_paid;

		dialog.fields_dict.amount_summary_display.$wrapper
			.find('#balance_to_pay_display')
			.html(`Balance to Pay: ₹${balance.toFixed(2)}`);
	};
	frappe.realtime.on('editable_table_changed', () => {
		update_balance();
	});

	dialog.fields_dict.payments.grid.wrapper.on('input', function () {
		update_balance();
	});

	dialog.fields_dict.payments.grid.wrapper.on('click', '.grid-add-row', () => {
		setTimeout(() => {
			const grid = dialog.fields_dict.payments.grid;
			const all_rows = grid.get_data();

			let paid_so_far = 0;
			all_rows.forEach(row => {
				paid_so_far += flt(row.amount || 0);
			});

			const remaining = flt(frm.doc.outstanding_amount) - paid_so_far;

			const last_row = grid.grid_rows[grid.grid_rows.length - 1];
			if (last_row && remaining > 0) {
				frappe.model.set_value(
					last_row.doc.doctype,
					last_row.doc.name,
					'amount',
					remaining
				);
			}

			update_balance();
		}, 100);
	});
	update_balance();
}

File: e_mart/e_mart/custom_scripts/sales_invoice/sales_invoice.py
import frappe
from frappe.utils import flt,add_months,nowdate,get_first_day,get_last_day,add_days,getdate
from frappe.model.mapper import get_mapped_doc
from datetime import datetime
from frappe.utils import flt
from frappe.utils import get_url_to_form
from frappe.utils import get_datetime


def on_submit(doc, method=None):
	create_scrap_stock_entry(doc, method)
	create_buyback_journal_entry(doc, method)
	create_demo_tasks_on_submit(doc, method)
	update_monthly_commission_log(doc, method)

def validate_buyback_fields(doc, method=None):
	"""
	Triggered on validation of Sales Invoice.

	1. Calculates amount for each Buyback Item row.
	2. Sums up all row amounts into buyback_amount.
	3. Adjusts the outstanding_amount, rounded total and grand total if is_buyback check box is checked
	"""

	total = 0

	# 1. Calculate amount per row
	if doc.buyback_items:
		for row in doc.buyback_items:
			row.amount = (row.qty or 0) * (row.rate or 0)
			total += row.amount

	# 2. Set total buyback amount
	doc.buyback_amount = total

	# 3.Calculate outstanding amount
	grand_total = (doc.total or 0) + (doc.total_taxes_and_charges or 0)
	if doc.is_buyback and doc.buyback_amount:
		grand_total -= doc.buyback_amount
	doc.outstanding_amount = round(grand_total)

	# 4. Set rounded_total and Grand Total
	grand_total = (doc.total or 0) + (doc.total_taxes_and_charges or 0)
	if doc.is_buyback and doc.buyback_amount:
		grand_total -= doc.buyback_amount
	doc.grand_total = grand_total
	doc.outstanding_amount = round(grand_total)
	doc.rounded_total = round(grand_total)

def create_scrap_stock_entry(doc, method):
	"""
	On submission of Sales Invoice:
	Creates a Stock Entry to move Buyback Items to Scrap Warehouse
	(Scrap warehouse is fetched from E-mart Settings).
	"""

	if not doc.buyback_items:
		return

	scrap_warehouse = frappe.db.get_single_value("E-mart Settings", "scrap_warehouse")
	if not scrap_warehouse:
		frappe.throw("Please set the Scrap Warehouse in E-mart Settings.")

	stock_entry = frappe.new_doc("Stock Entry")
	stock_entry.purpose = "Material Receipt"
	stock_entry.stock_entry_type = "Material Receipt"
	stock_entry.company = doc.company
	stock_entry.to_warehouse = scrap_warehouse
	stock_entry.sales_invoice_no = doc.name  # Optional

	for row in doc.buyback_items:
		if row.item and flt(row.qty) > 0:
			stock_entry.append("items", {
				"item_code": row.item,
				"qty": row.qty,
				"uom": "Nos",  # adjust as needed
				"t_warehouse": scrap_warehouse,
				"basic_rate": row.rate
			})

	stock_entry.insert()
	stock_entry.submit()
	frappe.msgprint(f' Scrap Stock Entry Created: <a href="{frappe.utils.get_url_to_form(stock_entry.doctype, stock_entry.name)}" target="_blank"><b>{stock_entry.name}</b></a>',alert=True,indicator='green')

def update_emi_amount(doc, method):
	"""
	Generate the emi amount after deducting the down payment
	"""
	for item in doc.items:
		 down_payment = item.down_payment or 0
		 amount = item.amount or 0
		 item.emi_amount = amount - down_payment
	
@frappe.whitelist()
def create_finance_invoice(sales_invoice_name):
	'''
	Create Finance Invoice from Sales Invoice
	'''
	sales_invoice = frappe.get_doc("Sales Invoice", sales_invoice_name)

	finance_invoice = frappe.new_doc("Finance Invoice")
	finance_invoice.customer = sales_invoice.emi_provider
	finance_invoice.posting_date = frappe.utils.nowdate()
	finance_invoice.total_amount = sales_invoice.total
	finance_invoice.actual_customer = sales_invoice.customer
	finance_invoice.total_taxes_and_charges = sales_invoice.total_taxes_and_charges

	for item in sales_invoice.items:
		finance_invoice.append("items", {
			"actual_item": item.item_code,
			"qty": item.qty,
			"rate": item.rate,
			"amount": item.amount
		})

	for tax in sales_invoice.taxes:
		finance_invoice.append("sales_taxes_and_charges", {
			"charge_type": tax.charge_type,
			"account_head": tax.account_head,
			"description": tax.description,
			"rate": tax.rate,
			"tax_amount": tax.tax_amount,
			"total": tax.total,
			"tax_amount_after_discount_amount": tax.tax_amount_after_discount_amount
		})

	finance_invoice.save(ignore_permissions=True)

	return finance_invoice.name

@frappe.whitelist()
def make_down_payment_entry(source_name, target_doc=None):
	def set_missing_values(source, target):
		"""
		Create a Payment Entry from a Sales Invoice for down payment.
		Replaces total payment with down payment amount and map hte details.
		
		"""
		down_payment = source.down_payment_amount
		target.payment_type = "Receive"
		target.party_type = "Customer"
		target.paid_amount = down_payment
		target.received_amount = down_payment

		paid_from = frappe.get_value("Company", source.company, "default_receivable_account")
		paid_to = frappe.get_value("Company", source.company, "default_bank_account")

		target.paid_from = paid_from
		target.paid_to = paid_to
		target.posting_date = frappe.utils.nowdate()

		paid_from_currency = frappe.get_value("Account", paid_from, "account_currency")
		paid_to_currency = frappe.get_value("Account", paid_to, "account_currency")

		target.paid_from_account_currency = paid_from_currency
		target.paid_to_account_currency = paid_to_currency

		target.append("references", {
			"reference_doctype": "Sales Invoice",
			"reference_name": source.name,
			"total_amount": source.rounded_total,
			"outstanding_amount": source.outstanding_amount,
			"allocated_amount": down_payment
		})

	doc = get_mapped_doc(
		"Sales Invoice",
		source_name,
		{
			"Sales Invoice": {
				"doctype": "Payment Entry",
				"field_map": {
					"customer": "party",
					"customer_name": "party_name",
					"company": "company"
				}
			}
		},
		target_doc,
		set_missing_values
	)
	return doc
	
def create_buyback_journal_entry(doc, method):
	'''
	Create Buyback Journal entry for Buyback amount from Sales invoice.
	'''
	if not doc.is_buyback or not doc.buyback_amount or doc.buyback_amount <= 0:
		return

	customer_account = doc.debit_to
	if not customer_account:
		frappe.throw(f"Customer receivable account (debit_to) not found in Sales Invoice {doc.name}")

	company = doc.company or frappe.db.get_value("Sales Invoice", doc.name, "company")
	if not company:
		frappe.throw(f"Company not found for Sales Invoice {doc.name}")

	buyback_account = frappe.db.get_single_value("E-mart Settings", "buyback_posting_account")
	if not buyback_account:
		frappe.throw("Please set 'Buyback Posting Account' in E-mart Settings.")

	# Create Journal Entry
	journal_entry = frappe.new_doc("Journal Entry")
	journal_entry.voucher_type = "Credit Note"
	journal_entry.posting_date = nowdate()
	journal_entry.company = company
	journal_entry.remark = f"Buyback adjustment for Sales Invoice {doc.name}"

	# Credit Customer (receivable)
	journal_entry.append("accounts", {
		"account": customer_account,
		"party_type": "Customer",
		"party": doc.customer,
		"credit_in_account_currency": doc.buyback_amount,
		"reference_type": "Sales Invoice",
		"reference_name": doc.name
	})

	# Dedit Buyback Posting Account
	journal_entry.append("accounts", {
		"account": buyback_account,
		"debit_in_account_currency": doc.buyback_amount
	})

	journal_entry.insert(ignore_permissions=True)
	journal_entry.submit()

	# Optional: Save Journal Entry reference if field exists
	if frappe.get_meta(doc.doctype).has_field("buyback_journal_entry"):
		doc.db_set("buyback_journal_entry", journal_entry.name)

	frappe.msgprint(
		f'Buyback Journal Entry Created: <a href="{frappe.utils.get_url_to_form("Journal Entry", journal_entry.name)}" target="_blank"><b>{journal_entry.name}</b></a>',alert=True,indicator='green')

def create_demo_tasks_on_submit(doc, method=None):
	"""
	Create a Task for each Sales Invoice Item where is_demo_reqd is checked.
	Subject, Task Type, Minimal Duration, Escalation Duration come from E-mart Settings.
	Maps invoice_date, invoice_reference, customer, exp_start_date, and end_date into Task.
	"""

	settings = frappe.get_single("E-mart Settings")
	task_subject_template = settings.subject or "Demo Required - {item_name} [{invoice_name}]"
	task_type = settings.task_type or "Demo"
	minimal_duration = settings.minimal_duration or 0
	escalation_duration = settings.escalation_duration or 0

	for item in doc.items:
		if item.get("is_demo_reqd"):
			subject = task_subject_template.format(
				item_name=item.item_name,
				invoice_name=doc.name
			)

			exp_start_date = add_days(doc.posting_date, minimal_duration)
			end_date = add_days(doc.posting_date, escalation_duration)

			task = frappe.new_doc("Task")
			task.subject = subject
			task.type = task_type
			task.status = "Open"
			task.reference_type = "Sales Invoice"
			task.reference_name = doc.name
			task.description = (
				f"Demo required for Item: {item.item_name} ({item.item_code})\n"
				f"Qty: {item.qty}\n"
				f"Customer: {doc.customer}\n"
			)
			
			task.invoice_date = doc.posting_date
			task.invoice_reference = doc.name
			task.customer = doc.customer
			task.exp_start_date = exp_start_date
			task.exp_end_date = end_date

			task.insert(ignore_permissions=True)

def update_monthly_commission_log(doc, method):
	"""
	Create or update Monthly Commission Log for each Sales Person in the Sales Invoice.
	Logs incentives and invoice details for the respective employee and month.
	"""
	invoice_date = datetime.strptime(doc.posting_date, "%Y-%m-%d").date()
	month_start = get_first_day(invoice_date)
	month_end = get_last_day(invoice_date)
	month_name = invoice_date.strftime('%B')

	for sales_team_member in doc.sales_team:
		sales_person = sales_team_member.sales_person
		incentives = sales_team_member.incentive or 0

		# Get the Employee linked to Sales Person
		employee = frappe.db.get_value("Sales Person", sales_person, "employee")
		if not employee:
			frappe.log_error(
				f"Sales Person {sales_person} has no linked Employee.",
				"Monthly Commission Log"
			)
			continue

		# Check if Monthly Commission Log exists
		log_name = frappe.db.exists(
			"Monthly Commission Log",
			{
				"employee": employee,
				"log_month": month_name,
				"start_date": month_start,
				"end_date": month_end
			}
		)

		if log_name:
			log = frappe.get_doc("Monthly Commission Log", log_name)
		else:
			log = frappe.new_doc("Monthly Commission Log")
			log.employee = employee
			log.log_month = month_name
			log.start_date = month_start
			log.end_date = month_end

		# Append detail row
		log.append("monthly_commission_log", {
			"sales_invoice": doc.name,
			"date": invoice_date,
			"total_amount": doc.base_grand_total,
			"incentives": incentives
		})

		log.save(ignore_permissions=True)
		link = get_url_to_form("Monthly Commission Log", log.name)
		frappe.msgprint(
			f'Monthly Commission Log Created/Updated: <a href="{link}" target="_blank"><b>{log.name}</b></a>',
			alert=True,
			indicator='green'
		)

def calculate_total_expense(doc, method):
	"""
	Calculate and set the total of all sales_expenses in the document.
	"""
	total = 0

	for row in doc.get("sales_expenses", []):
		total += flt(row.amount or 0)
	doc.total_expense = total

# calculate incentives based on commission rate and allocated percentage
def map_commission_to_sales_team(doc, method):
	"""Fetch total_commission_rate to each sales_team row as allocated_amount and compute incentives."""
	commission_rate = doc.total_commission_rate or 0

	for row in doc.sales_team:
		row.total_commission_rate = commission_rate
		allocated_percentage = row.allocated_percentage or 0
		row.incentive = round((commission_rate * allocated_percentage) / 100, 2)

def calculate_profit_for_commission(doc, method):
	"""Calculates profit for commission for each item by subtracting its Sales Expense Contribution from the total expense."""
	total_expense = flt(doc.total_expense or 0)

	for item in doc.items:
		contribution = flt(item.sales_expense_contribution or 0)
		profit = total_expense - contribution
		item.profit_for_commission = profit

def generate_emi_schedule(emi_doc, start_date, total_amount, no_of_installments):
	"""
	Adds EMI Schedule rows to the EMI Information document and returns the closing date.
	"""
	amount_per_installment = flt(total_amount) / no_of_installments
	start_date = getdate(start_date)
	last_date = None

	for i in range(no_of_installments):
		installment_date = add_months(start_date, i)
		last_date = installment_date 

		emi_doc.append("emi_schedule", {
			"no": i + 1,
			"date": installment_date,
			"amount": round(amount_per_installment, 2)
		})

	return last_date

@frappe.whitelist()
def create_emi_information(doc, method):
	"""
	 Creates EMI Information for each item marked as EMI.
	"""
	for item in doc.items:
		if item.is_emi:
			if not all([item.emi_amount, item.emi_start_date, item.no_of_installment]):
				frappe.throw(f"Missing EMI details for item {item.item_code}")

			emi_doc = frappe.new_doc("EMI Information")
			emi_doc.customer = doc.customer
			emi_doc.item = item.item_code
			emi_doc.emi_amount = item.emi_amount
			emi_doc.down_payment = item.down_payment
			emi_doc.emi_provider = doc.emi_provider
			emi_doc.emi_start_date = item.emi_start_date
			emi_doc.no_of_installment = item.no_of_installment
			emi_doc.sales_invoice = doc.name

			# Generate EMI Schedule and get closing date
			closing_date = generate_emi_schedule(
				emi_doc,
				start_date=item.emi_start_date,
				total_amount=item.emi_amount,
				no_of_installments=item.no_of_installment
			)
			emi_doc.closing_date = closing_date
			emi_doc.insert(ignore_permissions=True)

def calculate_total_down_payment(doc, method):
	"""
	Calculates the total down payment amount from items and sets it in the Sales Invoice.
	"""
	total_down_payment = 0

	for item in doc.items:
		if item.down_payment:
			total_down_payment += item.down_payment

	doc.down_payment_amount = total_down_payment

def calculate_total_emi_amount(doc, method):
	"""
	Calculates the total emi amount from items and sets it in the Sales Invoice.
	"""
	total_emi_amount = 0

	for item in doc.items:
		if item.emi_amount:
			total_emi_amount += item.emi_amount

	doc.emi_amount = total_emi_amount

def get_valuation_rate(item_code, warehouse, posting_date, posting_time):
	"""
	Get the most recent valuation rate from Stock Ledger Entry (SLE).
	"""
	if not item_code or not warehouse:
		return 0

	posting_datetime = get_datetime(f"{posting_date} {posting_time}")

	sle = frappe.db.sql("""
		SELECT valuation_rate
		FROM `tabStock Ledger Entry`
		WHERE item_code = %s
		AND warehouse = %s
		AND TIMESTAMP(posting_date, posting_time) <= %s
		AND valuation_rate IS NOT NULL
		ORDER BY posting_date DESC, posting_time DESC, creation DESC
		LIMIT 1
	""", (item_code, warehouse, posting_datetime), as_dict=True)

	return sle[0].valuation_rate if sle else 0


def set_valuation_and_gross_profit(doc, method):
	"""
	Set valuation_rate and gross_profit for each item in Sales Invoice.
	"""
	for item in doc.items:
		if not item.item_code or not item.warehouse:
			continue

		# Fetch valuation rate from SLE
		valuation_rate = get_valuation_rate(
			item.item_code,
			item.warehouse,
			doc.posting_date,
			doc.posting_time
		)

		item.valuation_rate = valuation_rate or 0
		item.gross_profit = (item.amount or 0) - (valuation_rate or 0) * (item.qty or 0)


